diff -r aa447af2817b programs/setup/setup.in
--- a/programs/setup/setup.in	Fri Apr 08 19:41:58 2011 +0900
+++ b/programs/setup/setup.in	Fri Apr 08 19:55:04 2011 +0900
@@ -5,7 +5,7 @@
 # Provides:          ipsec
 # Required-Start:    $network $remote_fs $syslog $named
 # Required-Stop:     $syslog $remote_fs
-# Default-Start:
+# Default-Start:     2 3 4 5
 # Default-Stop:      0 1 6
 # Short-Description: Start Libreswan IPsec at boot time
 # Description:       Enable automatic key management for IPsec (KLIPS and NETKEY)
