--- /tmp/linux/Makefile	Fri Jan 12 10:38:19 2001
+++ linux/Makefile	Fri Jan 12 10:34:21 2001
@@ -265,6 +265,9 @@
 oldconfig: symlinks
 	$(CONFIG_SHELL) scripts/Configure -d arch/$(ARCH)/config.in
 
+oldconfig_nonint: symlinks
+	$(CONFIG_SHELL) scripts/Configure -D arch/$(ARCH)/config.in
+
 xconfig: symlinks
 	$(MAKE) -C scripts kconfig.tk
 	wish -f scripts/kconfig.tk
--- /tmp/linux/scripts/Configure	Sat Dec 30 21:16:13 2000
+++ linux/scripts/Configure	Fri Jan 12 10:36:51 2001
@@ -121,8 +121,14 @@
 	else
 		echo -n "$1"
 		[ -z "$3" ] && echo -n "(NEW) "
-		IFS='@' read ans || exit 1
-		[ -z "$ans" ] && ans=$2
+		if [ "$NONINTERACTIVE" = "y" ]; then
+			echo "$1" >> nodefaults
+			broken="y"
+			ans="n"
+		else
+			IFS='@' read ans || exit 1
+			[ -z "$ans" ] && ans=$2
+		fi
 	fi
 }
 
@@ -519,6 +525,7 @@
 CONFIG=.tmpconfig
 CONFIG_H=.tmpconfig.h
 trap "rm -f $CONFIG $CONFIG_H ; exit 1" 1 2
+trap "rm -f nodefaults ; exit 1" 1 2
 
 #
 # Make sure we start out with a clean slate.
@@ -533,11 +540,20 @@
 echo "#define AUTOCONF_INCLUDED" >> $CONFIG_H
 
 DEFAULT=""
+NONINTERACTIVE=""
+
 if [ "$1" = "-d" ] ; then
 	DEFAULT="-d"
 	shift
 fi
 
+if [ "$1" = "-D" ] ; then
+# non interactive oldconfig
+	DEFAULT="-d"
+	NONINTERACTIVE="y"
+	shift
+fi
+
 CONFIG_IN=./config.in
 if [ "$1" != "" ] ; then
 	CONFIG_IN=$1
@@ -581,4 +597,9 @@
 fi
 echo
 
+if [ "$broken" = "y" ] ; then
+ echo "The following defaults are missing:"
+ cat nodefaults
+ exit 1
+fi
 exit 0
