--- a/include/net/xfrm.h.orig	2006-12-27 11:34:51.000000000 -0500
+++ b/include/net/xfrm.h	2007-02-06 07:40:26.000000000 -0500
@@ -596,13 +596,19 @@
 
 extern void xfrm_dst_ifdown(struct dst_entry *dst, struct net_device *dev);
 
+typedef unsigned int xfrm_sec_unique_t;
+ 
 struct sec_path
 {
 	atomic_t		refcnt;
+ 	xfrm_sec_unique_t       ref;       /*reference to high-level policy*/
 	int			len;
 	struct xfrm_state	*xvec[XFRM_MAX_DEPTH];
 };
 
+struct ipcm_cookie;
+extern int ip_cmsg_send_ipsec(struct cmsghdr *cmsg, struct ipcm_cookie *ipc);
+ 
 static inline struct sec_path *
 secpath_get(struct sec_path *sp)
 {
